flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -327,23 +358,83 @@ private MemorySegment requestMemorySegment() {
 		return requestMemorySegment(UNKNOWN_CHANNEL);
 	}
 
-	@Nullable
-	private MemorySegment requestMemorySegmentFromGlobal() {
+	private boolean requestMemorySegmentFromGlobal() {
+		assert Thread.holdsLock(availableMemorySegments);
+
+		if (isRequestedSizeReached()) {
+			return false;
+		}
+
+		MemorySegment segment = networkBufferPool.requestMemorySegment();
+		if (segment != null) {
+			availableMemorySegments.add(segment);
+			numberOfRequestedMemorySegments++;
+			return true;
+		}
+		return false;
+	}
+
+	private boolean isRequestedSizeReached() {
+		return numberOfRequestedMemorySegments >= currentPoolSize;
+	}
```

Comment:
nit: there are still two places doing `numberOfRequestedMemorySegments > currentPoolSize`, but I guess they are off by one in the comparison and can not be easily migrated to `isRequestedSizeReached()`?

Reply:
I didn't find a good way, but extracted them into `hasExcessBuffers` for now.